README.rst
====================================
weetwit: A twitter suite for weechat
====================================

*author*
    Daniël Franke (@ainmosni) <daniel[at]ams-sec[dot]org>

Welcome to the wonderful world of weetwit, a plugin that will transform weechat
to a (soon-to-be) fully-featured twitter client.

Features
========

* Realtime streaming timeline.
* Realtime search.
* Trending support.
* Tweet length counter.
* Favourite support.
* Updating, replying and retweeting according to Twitter standards.
* Aware of twitter URL shortening.
* Tools to query users and status updates.
* Able to (un-)follow directly from weechat.
* Report spammers directly from weechat.
* Share updates with IRC.
* Conversation support


Installation
============

Before you can use it, you have to "create" a new application at twitter, the
reason for this is that I can't include the application keys without them
becoming public knowledge. You can create the application at `Twitter
<http://dev.twitter.com>`_. The application needs read/write access.

First we need to install the module:

Installation through pip, note that you must get the plugin seperately.::

    # pip install weetwit
    $ Copy the plugin to ~/.weechat/python

Installation from source::

    # cd /path/to/source/files
    # pip install tweepy
    # python setup.py install
    $ cp plugin/weetwit.py ~/.weechat/python

Configuration of weechat::

    ] /python load weetwit.py
    ] /set plugins.var.python.weetwit.access_token access_token_goes_here
    ] /set plugins.var.python.weetwit.access_token_secret access_token_secret_goes_here
    ] /set plugins.var.python.weetwit.consumer_key consumer_key_goes_here
    ] /set plugins.var.python.weetwit.consumer_secret consumer_secret_goes_here
    ] /python reload weetwit

You should now have a running weetwit.

Configuration parameters
========================

There are multiple configuration parameters you can alter, here is a short
description of them.

* **plugins.var.python.weetwit.show_in_current**: Show the timeline in the
  current window.
* **plugins.var.python.weetwit.current_while_away**: Show the timeline in the
  current window while away.
* **plugins.var.python.weetwit.current_while_detached**: Show the timeline in the
  current window while screen/tmux detached.
* **plugins.var.python.weetwit.current_color**: The color of the tweets in the
  current buffer.
* **plugins.var.python.weetwit.storage_dir**: The location of where all the
  weetwit related files will be kept.
* **plugins.var.python.weetwit.timelined_location**: The location of the
  timelined monitoring daemon.
* **plugins.var.python.weetwit.trend_woeid**: The woeid you want to see trends
  for, defaults to worldwide.
* **plugins.var.python.weetwit.nick_color**: The color of @names. Use
  'nick_color' if you want people colored uniquely.
* **plugins.var.python.weetwit.hash_color**: The color of #hashtags.
* **plugins.var.python.weetwit.mention_color**: The color that @people
  mentioned in tweets should have.
* **rt_style**: How RTs are displayed. 'postfix' will show the retweeter after
  the tweet like this (RT by @username) 'prefix' will show the retweeter before
  the tweet.
* **expand_urls**: Expand URLs, when this is on, it will show a preview of the
  URL before the t.co URL, if off it will only show the t.co url.


Bar items
=========

* **tweet_counter** Shows the amount of characters that are typed into the
  timeline buffer, it's aware of t.co URL shortening.

Weechat commands
================

Many commands take <status identification> as argument, this can either be the
ID of the status or a screen_name. In case of the screen_name, we will use the
ID of last status by screen_name. (Note: if screen_name hasn't showed up in your
timeline, this won't work.)

* \/tweet <status>
    - Update your status, this can be 140 characters long. URLs will be shortened 
      using t.co by twitter. You don't need this command on the dedicated
      timeline buffer.

* \/tinfo <status identification>
    - Shows more detailed information about a status update.

* \/treply [.]<status identification> <message>
    - Replies to the relevant status update, this will always start with the
      @screen_name of the person the status belonged to, if the identification is
      prefixed with a dot, a dot will be prepended to the message so that your
      other followers will see it as well.

* \/tconversation <status identification>
    - Displays the conversation leading up this status update.

* \/tfavorite <status identification>
    - Favourites a status update.

* \/tunfavorite <status identification>
    - Removes a tweet from your favourites.

* \/tfavorites
    - Shows your favourited tweets.


* \/retweet <status identification> [message]
    - Retweets the relevant status update, if [message] is present this will
      prepended to the retweet.

* \/tsearch <keywords>
    - Opens a new buffer with a realtime search of <keywords>, you can only
      have a limited amount open of these at one time, opening more might stop
      already existing searches.

* \/tshare <status identification>
    - Shares the relevant status update with the current IRC channel.

* \/twhois <screen_name>
    - Displays information about screen_name.

* \/tfollow <screen_name>
    - Follows screen_name.

* \/tunfollow <screen_name>
    - Unfollows screen_name.

* \/trending [woeid]
    - Displays what's trending in the location represented by [woeid].
      If no woeid present it uses the woeid set at
      plugins.var.python.weetwit.trend_woeid.

* \/travail
    - Displays woeids of available trend locations.

* \/treport [--yes] <screen_name>
    - Reports <screen_name> for spam. If --yes isn't added, the user won't be
      reported for spam.

FAQ
===

    :Q: Why does your script spawn an extra python process?
    :A: Because weechat doesn't support background threads, and blocks on long
        running operations, this process is what monitors your timeline.


    :Q: I don't want those ugly STATUSIDs in my weetwit buffer.
    :A: Add a filter like this: "/filter add statusid python.timeline * \\[#STATUSID:"
        now you can toggle between them hidden and visible, depending on your
        needs.


CHANGELOG.rst
Changelog
---------

0.10.0 (2014-02-27)
++++++++++++++++++

* Fix the /trending command.
* Add more granular control over when the timeline appears in your current
  buffer.

0.9.1 (2013-06-17)
++++++++++++++++++

* Fix to work with tweepy 2.0, needed because twitter killed the 1.0 API.
* Because of the API upgrade, weetwit now requires tweepy 2.0.
* Yes, I'm still alive and will fix breakages. Features will come slowly
  though, if you want it badly enough, either ask me through twitter/IRC or
  shoot me a pull request.

0.9.0 (Unreleased)
++++++++++++++++++

* Added STATUSID markers to the conversation view.
* Added /tfavorites to show your favourites.
* Added /tfavorite to add a tweet to favourites.
* Added /tunfavorite to remove a tweet from your favourites.

0.8.0 (2012-04-12)
++++++++++++++++++

* Added a length counter that can be added to a bar.
* Made it possible to tweet from the timeline buffer.
* Handled a stacktrace better when inputting a wrong woeid.
* Added a command to report spam.
* Added the twitter client to /tinfo.

0.7.0 (2012-04-03)
++++++++++++++++++++

* Added realtime search.
* Removed logging.
* Added command to show available trending locations.
* Added command to see what's trending.

0.6.2 (2012-03-21)
++++++++++++++++++

* Changed displayed timestamps to when the tweet was created.
* Added the user's own screen_name to the nicklist.


libweetwit/twitter.py
#!/usr/bin/env python
#coding: utf-8
#
# File Name: twitter.py
#
# Description:
#
# Creation Date: 2012-02-22
#
# Last Modified: 2013-06-17 23:42
#
# Created By: Daniël Franke <daniel@ams-sec.org>
#

import re

from collections import defaultdict

from tweepy.parsers import ModelParser
from tweepy import OAuthHandler, TweepError

from libweetwit.wtmodelfactory import wtModelFactory
from libweetwit.db import DB
from libweetwit.exceptions import TwitterError
from libweetwit.api import weetwitAPI as API


class Twitter(object):
    """A class that does all interactions with twitter."""
    def __init__(self, storage_dir=False, db=False):
        """Initialise the API."""
        if not db and not storage_dir:
            raise TypeError(
                "Twitter() needs either a storage_dir or a db argument."
            )

        if not db:
            db = DB(storage_dir)
        self.db = db
        ck = self.db.get_config('consumer_key')
        cs = self.db.get_config('consumer_secret')
        at = self.db.get_config('access_token')
        ats = self.db.get_config('access_token_secret')
        mf = wtModelFactory()
        pr = ModelParser(model_factory=mf)
        self.auth = OAuthHandler(ck, cs)
        self.auth.set_access_token(at, ats)
        self.api = API(self.auth, parser=pr)
        try:
            self.api.me().name
        except TweepError as error:
            raise TwitterError("Could not connect to Twitter: %s" % error)
        except TypeError as error:
            raise TwitterError("Your keys haven't been set correctly!")

    def update_status(self, message, reply_id=False):
        """Posts text to twitter."""
        if self.__is_sane(message):
            try:
                self.api.update_status(message, in_reply_to_status_id=reply_id)
            except TweepError as error:
                raise TwitterError("Failed to post status: %s" % error)
            return "Status updated."
        else:
            raise TwitterError("Status too long!")

    def get_tweet(self, identification):
        """Return a tweet from either an integer or cached screen name."""
        tid = False
        try:
            int(identification)
            tid = identification
        except ValueError:
            identification = identification.lstrip("@")
            tid = self.db.get_last_tid(identification)
            if not tid:
                raise TwitterError("ID %s not cached." % identification)
        try:
            return self.api.get_status(tid, include_entities=True)
        except TweepError as error:
            raise TwitterError("Failed to get tweet: %s" % error)

    def get_user(self, user):
        """Returns the requested user."""
        try:
            user = self.api.get_user(user, include_entities=True)
        except TweepError as error:
            raise TwitterError("Failed to get user: %s" % error)
        return user

    def get_followed(self):
        """Returns an array of screen_names that you follow."""
        try:
            followed = [u.screen_name for u in self.api.friends()]
        except TweepError as error:
            raise TwitterError("Faled to get followed: %s" % error)
        return followed

    def get_trend_places(self):
        """
        Returns a dict of dicts, first by country then by place.
        Every country has a special 'woeid' key that holds the woeid of the
        country itself and all the places of the country pointing to their
        respective woeids.
        A special exception is the 'Worldwide' "country" which only has a
        woeid entry.
        """
        try:
            trend_places = self.api.trends_available()
        except TweepError as error:
            raise TwitterError("Falied to get available places: %s." % error)

        places = defaultdict(dict)
        for place in trend_places:
            if not place['country']:
                places['Worldwide'] = {'woeid': place['woeid']}
            else:
                if place['country'] == place['name']:
                    places[place['country']]['woeid'] = place['woeid']
                else:
                    places[place['country']][place['name']] = place['woeid']
        return places

    def get_trends(self, woeid):
        """
        Gets the current trends for the location represented by woeid.
        Returns a list of trends with element 0 representing the location name.
        """
        try:
            trend_response = self.api.trends_place(woeid)
        except TweepError as error:
            raise TwitterError("Failed to get trends: %s" % error)

        trends = []
        trends.append(trend_response[0]['locations'][0]['name'])
        for trend in trend_response[0]['trends']:
            trends.append(trend['name'])

        return trends

    def get_favorites(self):
        """Get your favourite tweets."""
        try:
            favorites = self.api.favorites(include_entities=True)
        except TweepError as error:
            raise TwitterError("Failed to get favourites: %s" % error)
        return favorites

    def status_count(self, message):
        """
        Replaces the URLs and returns the length how twitter would count it.
        """
        message = self.__replace_urls(message)
        return(len(message))

    def __is_sane(self, message):
        """Does sanity checks to see if the status is valid."""
        if self.status_count(message) > 140:
            return False
        return True

    def __replace_urls(self, message):
        """Replace URLs with placeholders, 20 for http URLs, 21 for https."""
        # regexes to match URLs
        octet = r'(?:2(?:[0-4]\d|5[0-5])|1\d\d|\d{1,2})'
        ip_addr = r'%s(?:\.%s){3}' % (octet, octet)
        # Base domain regex off RFC 1034 and 1738
        label = r'[0-9a-z][-0-9a-z]*[0-9a-z]?'
        domain = r'%s(?:\.%s)*\.[a-z][-0-9a-z]*[a-z]?' % (label, label)
        url_re = re.compile(r'(\w+://(?:%s|%s)(?::\d+)?(?:/[^\])>\s]*)?)' %
                            (domain, ip_addr), re.I)

        new_message = message

        for url in url_re.findall(message):
            short_url = 'x' * 20
            if url.startswith('https'):
                short_url = 'x' * 21
            new_message = new_message.replace(url, short_url)

        return new_message
